File: aksharallm/portal/static/js/diffusion.js
/* The Diffusion tab: the second paradigm, which is far easier to watch than to read about.
 *
 * The whole tab is built around one claim — that the *order* a diffusion model decides
 * positions in is the interesting thing, and it is invisible in the finished text. So the
 * server returns every intermediate state of a generation and this module animates them.
 * Everything else here (the corruption preview, infilling, the measurement panel) exists to
 * make that one animation mean something.
 *
 * Runs inline on the Playground's resident model, like the Interp tab, so it inherits the
 * device policy: while a run is training the model is on the CPU and the status line says so.
 */
import { $, $$, api, escHtml, flash, fmt, post } from './core.js';
import { registerTab } from './router.js';

const diff = { loaded: false, ckpt: null, trace: null, at: 0, timer: null, prefixLen: 0 };

function status(text, kind = '') {
  const el = $('#diff-status');
  el.textContent = text;
  el.className = `panel-note${kind ? ` ${kind}` : ''}`;
}

/* ---- checkpoints ---------------------------------------------------------------------- */

async function loadCheckpoints() {
  const res = await api(`/api/diffusion${diff.ckpt ? `?checkpoint=${encodeURIComponent(diff.ckpt)}` : ''}`);
  const sel = $('#diff-ckpt');
  const cks = res.checkpoints || [];
  /* Non-diffusion checkpoints are listed and disabled rather than hidden: a picker that
   * silently drops every checkpoint on a machine that has not trained one looks broken. */
  sel.innerHTML = cks.map((c) => `<option value="${escHtml(c.rel)}"${c.diffusion ? '' : ' disabled'}>`
    + `${escHtml(c.rel)}${c.diffusion ? '' : ' — autoregressive'}</option>`).join('');
  const first = cks.find((c) => c.diffusion);
  $('#diff-empty').hidden = !!first;
  if (first) {
    diff.ckpt = (cks.find((c) => c.rel === diff.ckpt && c.diffusion) || first).rel;
    sel.value = diff.ckpt;
  } else {
    diff.ckpt = null;
  }
  const cur = res.current;
  $('#diff-current').textContent = cur && !cur.error
    ? `[MASK] = id ${cur.mask_token_id} of ${fmt.int(cur.vocab_size)} · context ${fmt.int(cur.max_seq_len)} · on the ${cur.device}`
    : '';
  for (const id of ['#diff-run', '#diff-infill-run', '#diff-corrupt-run', '#diff-elbo', '#diff-byt']) {
    $(id).disabled = !diff.ckpt;
  }
  return res;
}

/* ---- 1. the forward process --------------------------------------------------------- */

async function runCorrupt() {
  if (!diff.ckpt) return;
  const t = Number($('#diff-t').value) / 100;
  $('#diff-t-label').textContent = `${Math.round(t * 100)}%`;
  try {
    const d = await post('/api/diffusion/corrupt', {
      checkpoint: diff.ckpt, text: $('#diff-corrupt-text').value,
      t, seed: Math.floor(Math.random() * 1e6),
    });
    $('#diff-corrupt-out').innerHTML = d.tokens.map((tok, i) => (d.masked[i]
      ? `<span class="cell masked" title="${escHtml(tok)} — the model has to put this back">▁</span>`
      : `<span class="cell">${escHtml(tok)}</span>`)).join('');
    $('#diff-weight').innerHTML =
      `${d.n_masked} of ${d.n_tokens} tokens hidden (asked for ${fmt.pct(d.t, 0)}, got `
      + `${fmt.pct(d.realised, 0)} — it is a coin per position, not a quota). Each one's `
      + `cross-entropy is multiplied by <strong>1/t = ${d.weight}</strong>, so this sequence `
      + `counts the same as one masked at 90%.`;
  } catch (err) {
    status(err.message, 'err');
  }
}

/* ---- 2. generation, and the animation ------------------------------------------------ */

function drawTrace(i) {
  if (!diff.trace) return;
  const steps = diff.trace.steps;
  diff.at = Math.max(0, Math.min(i, steps.length - 1));
  const st = steps[diff.at];
  const fresh = new Map();
  st.committed.forEach((p, k) => fresh.set(p, st.confidence ? st.confidence[k] : 1));

  $('#diff-trace').innerHTML = st.cells.map((text, pos) => {
    if (pos < diff.trace.prefix_len) {
      return `<span class="cell given">${escHtml(text || '')}</span>`;
    }
    if (text == null) return '<span class="cell masked">▁</span>';
    if (fresh.has(pos)) {
      /* Opacity is the confidence: a wall of pale cells on the first step is a model
       * guessing, and the same wall going dark by step three is one that has found its
       * skeleton. No number could show that as fast. */
      const w = Math.max(0.25, Math.min(fresh.get(pos), 1));
      return `<span class="cell fresh" style="--w:${w}" title="committed at step ${st.step}, `
        + `confidence ${fmt.num(fresh.get(pos), 2)}">${escHtml(text)}</span>`;
    }
    return `<span class="cell done">${escHtml(text)}</span>`;
  }).join('');

  $('#diff-scrub').value = String(diff.at);
  $('#diff-step-label').textContent = st.step === 0
    ? 'step 0 — everything blank'
    : `step ${st.step} of ${steps.length - 1} · ${st.committed.length} committed · ${st.remaining} left`;
}

function stopPlay() {
  if (diff.timer) clearInterval(diff.timer);
  diff.timer = null;
  $('#diff-play').textContent = '▶ play';
}

function play() {
  if (diff.timer) { stopPlay(); return; }
  if (!diff.trace) return;
  if (diff.at >= diff.trace.steps.length - 1) drawTrace(0);
  $('#diff-play').textContent = '❚❚ pause';
  diff.timer = setInterval(() => {
    if (diff.at >= diff.trace.steps.length - 1) { stopPlay(); return; }
    drawTrace(diff.at + 1);
  }, 420);
}

async function runGenerate() {
  if (!diff.ckpt) return;
  stopPlay();
  status('denoising…');
  $('#diff-run').disabled = true;
  try {
    const d = await post('/api/diffusion/generate', {
      checkpoint: diff.ckpt,
      prompt: $('#diff-prompt').value,
      length: Number($('#diff-length').value),
      steps: Number($('#diff-steps').value),
      temperature: Number($('#diff-temp').value),
      remask: $('#diff-remask').value,
    });
    diff.trace = d;
    $('#diff-player').hidden = false;
    $('#diff-scrub').max = String(d.steps.length - 1);
    drawTrace(0);
    $('#diff-cost').innerHTML =
      `<strong>${d.passes} forward passes</strong> for ${fmt.int($('#diff-length').value)} `
      + `tokens — ${d.tokens_per_pass} tokens per pass, where an autoregressive model needs `
      + `exactly one pass per token. ${fmt.dur(d.elapsed_s)} on the ${d.device}. `
      + `<em>But</em> each pass here is over the whole sequence with no cache, so this is a `
      + `trade, not a free win.`;
    status(`generated on the ${d.device}`);
    play();
  } catch (err) {
    status(err.message, 'err');
    flash(err.message, 'error');
  } finally {
    $('#diff-run').disabled = false;
  }
}

/* ---- 3. infilling --------------------------------------------------------------------- */

async function runInfill() {
  if (!diff.ckpt) return;
  status('filling the gap…');
  try {
    const d = await post('/api/diffusion/infill', {
      checkpoint: diff.ckpt,
      prefix: $('#diff-prefix').value,
      suffix: $('#diff-suffix').value,
      length: Number($('#diff-mid-len').value),
    });
    $('#diff-infill-out').innerHTML = `${escHtml(d.prefix)} `
      + `<mark>${escHtml(d.middle)}</mark> ${escHtml(d.suffix)}`;
    status(`filled on the ${d.device}`);
  } catch (err) {
    status(err.message, 'err');
  }
}

/* ---- 4. measurement ------------------------------------------------------------------- */

async function measure(kind) {
  if (!diff.ckpt) return;
  status(kind === 'elbo' ? 'measuring the bound…' : 'measuring across mask rates…');
  $('#diff-measure').innerHTML = '<p class="hint">running…</p>';
  try {
    const d = await post('/api/diffusion/measure', { checkpoint: diff.ckpt, kind });
    if (d.kind === 'elbo') {
      $('#diff-measure').innerHTML = `<div class="tiles">
        <div class="tile"><div class="tile-value">${fmt.num(d.nelbo, 4)}</div>
          <div class="tile-label">NELBO, nats/token</div>
          <div class="tile-note">an upper bound on the true NLL</div></div>
        <div class="tile"><div class="tile-value">&le; ${fmt.num(d.ppl_upper_bound, 2)}</div>
          <div class="tile-label">perplexity bound</div>
          <div class="tile-note">not an AR perplexity — note the &le;</div></div>
        <div class="tile"><div class="tile-value">${fmt.num(d.ce_masked, 3)}</div>
          <div class="tile-label">ce on masked positions</div>
          <div class="tile-note">unweighted; the number with a meaning</div></div>
      </div>
      <p class="hint">A quick reading over ${d.batches} batches. For a number to write down:
        <code>python -m aksharallm.diffusion ${escHtml(diff.ckpt)} elbo --batches 20</code></p>`;
      return;
    }
    const rows = d.rows || [];
    const max = Math.max(...rows.map((r) => r.ce_masked), 0.001);
    $('#diff-measure').innerHTML = '<div class="byt">' + rows.map((r) =>
      `<div class="byt-row"><span class="byt-label">${fmt.pct(r.t, 0)}</span>`
      + `<span class="byt-bar"><i style="width:${(r.ce_masked / max) * 100}%"></i></span>`
      + `<span class="byt-val">${fmt.num(r.ce_masked, 3)}</span></div>`).join('') + '</div>';
  } catch (err) {
    $('#diff-measure').innerHTML = `<p class="hint">${escHtml(err.message)}</p>`;
    status(err.message, 'err');
  }
}

/* ---- wiring ---------------------------------------------------------------------------- */

async function openDiffTab() {
  if (diff.loaded) return;
  diff.loaded = true;
  try {
    await loadCheckpoints();
    $('#diff-ckpt').addEventListener('change', async (e) => {
      diff.ckpt = e.target.value;
      await loadCheckpoints();
    });
    $('#diff-t').addEventListener('input', runCorrupt);
    $('#diff-corrupt-text').addEventListener('change', runCorrupt);
    $('#diff-corrupt-run').addEventListener('click', runCorrupt);
    $('#diff-run').addEventListener('click', runGenerate);
    $('#diff-play').addEventListener('click', play);
    $('#diff-scrub').addEventListener('input', (e) => { stopPlay(); drawTrace(Number(e.target.value)); });
    $('#diff-infill-run').addEventListener('click', runInfill);
    $('#diff-elbo').addEventListener('click', () => measure('elbo'));
    $('#diff-byt').addEventListener('click', () => measure('by-t'));
    if (diff.ckpt) runCorrupt();
  } catch (err) {
    status(err.message, 'err');
  }
}

registerTab('diff', { open: openDiffTab, leave: stopPlay });

File: aksharallm/portal/diffusion.py
"""The Diffusion tab's back end: watch a sentence resolve out of noise.

Shape of this one, and why
--------------------------
Like `interp`, everything here runs **inline** on the Playground's resident model rather
than as a subprocess job. A denoising run at Phase-1 scale is `steps` forward passes over a
few hundred tokens — well under a second on the card and a few seconds on the CPU — so a job
runner with a pid file and a log tail would be more machinery than the work. Reusing
`Playground.engine` also means this tab inherits the device policy, the idle unload and the
generation lock, so opening it can never cost somebody a training run.

The one thing here that touches no model at all is `corrupt_preview`, and it is the panel a
newcomer should meet first: it applies the *forward* process to text they typed and shows
what the model is asked to reconstruct. The whole objective is legible before anything is
generated, which is the same argument the Finetune tab's memory-budget table makes.

Read with: docs/19-diffusion.md -- the chapter this implements; it ends with the order to read
these files in.
"""

from __future__ import annotations

import time
from pathlib import Path

import torch

from ..diffusion.corrupt import corrupt, sample_t
from ..diffusion.evaluate import elbo, loss_by_t
from ..diffusion.generate import DiffusionError, diffusion_generate, infill
from ..infer.checkpoints import InferError

#: Ceilings, so a browser cannot ask for a denoising run that holds the model for a minute.
#: Generation is `steps` passes over the WHOLE sequence — there is no cache to make a longer
#: one cheap — so both numbers matter and neither is generous.
MAX_LENGTH = 256
MAX_STEPS = 128
#: Measurement is bounded the same way. The full curve is a CLI job.
MAX_BATCHES = 20


class Diffusion:
    """Everything the Diffusion tab can ask for, against the resident model."""

    def __init__(self, playground, root: Path | None = None):
        self.playground = playground
        self.root = Path(root) if root else Path.cwd()

    # ---- shared -----------------------------------------------------------------------

    def _loaded(self, ckpt_id: str):
        loaded = self.playground.engine.load(ckpt_id)
        if not loaded.model.cfg.is_diffusion:
            raise InferError(
                f"{loaded.info.rel} is an autoregressive checkpoint — causal attention and no "
                "[MASK] token, so it has never seen a masked sequence and unmasking it would "
                "return noise. Train configs/tiny-diffusion.yaml (the Dashboard's Start "
                "button) and come back. See docs/19.")
        return loaded

    def _cells(self, loaded, ids, mask_id: int) -> list:
        """Token ids -> one display string per position, or None where still masked."""
        out = []
        for i in ids:
            out.append(None if i == mask_id else loaded.tokenizer.decode([int(i)]))
        return out

    def overview(self, ckpt_id: str | None = None) -> dict:
        """What the tab needs before anything runs.

        Diffusion checkpoints are flagged here rather than filtered out: a picker that
        silently hides every checkpoint on a machine that has not trained one yet looks
        broken. Showing them all, greyed with a reason, is the honest empty state.
        """
        info = self.playground.overview()
        cks = []
        for c in info.get("checkpoints", []):
            cks.append({**c, "diffusion": bool(c.get("diffusion"))})
        out = {
            "checkpoints": cks,
            "any": any(c["diffusion"] for c in cks),
            "loaded": info.get("loaded"),
            "device": info.get("device"),
            "plan": info.get("plan"),
            "max_length": MAX_LENGTH, "max_steps": MAX_STEPS,
            "results": self.results(),
        }
        if ckpt_id:
            try:
                loaded = self._loaded(ckpt_id)
                out["current"] = {
                    "checkpoint": loaded.info.rel,
                    "mask_token_id": int(loaded.model.cfg.mask_token_id),
                    "vocab_size": loaded.model.cfg.vocab_size,
                    "max_seq_len": loaded.model.cfg.max_seq_len,
                    "device": loaded.device,
                }
            except (InferError, DiffusionError) as exc:
                out["current"] = {"error": str(exc)}
        return out

    # ---- the forward process, with no model in it ---------------------------------------

    def corrupt_preview(self, ckpt_id: str, text: str, t: float, seed: int = 0) -> dict:
        """Show what the model is *given* at mask rate `t`, and what it has to put back.

        No forward pass. This is the training objective's first half, applied to a sentence
        the reader typed — which is the cheapest possible way to understand what "noise"
        means for discrete tokens.
        """
        loaded = self._loaded(ckpt_id)
        mask_id = int(loaded.model.cfg.mask_token_id)
        ids = loaded.tokenizer.encode(text or "", bos=False)
        if not ids:
            raise InferError("type something to corrupt")
        ids = ids[:MAX_LENGTH]
        x = torch.tensor(ids, dtype=torch.long)[None, :]
        gen = torch.Generator().manual_seed(int(seed))
        rate = max(0.0, min(float(t), 1.0))
        c = corrupt(x, mask_id, torch.full((1,), rate), generator=gen)
        return {
            "t": rate,
            "realised": c.rate,
            "tokens": [loaded.tokenizer.decode([i]) for i in ids],
            "masked": c.masked[0].tolist(),
            # The 1/t weight, spelled out. At t = 0.05 one masked token counts for twenty;
            # seeing that number move as the slider does is the point of the panel.
            "weight": round(1.0 / max(rate, 1e-3), 1),
            "n_masked": int(c.masked.sum()),
            "n_tokens": len(ids),
        }

    # ---- generation ----------------------------------------------------------------------

    def generate(self, ckpt_id: str, *, prompt: str = "", length: int = 48,
                 steps: int = 16, temperature: float = 0.8, top_k: int = 50,
                 top_p: float = 0.95, remask: str = "low_confidence",
                 seed: int | None = None) -> dict:
        """Denoise a fresh sequence, and return every intermediate state.

        The trace is the whole reason this tab exists. The finished string says nothing
        about the order the positions were decided in, and that order is the model's
        reasoning made visible — the one thing about this paradigm that cannot be explained
        as well in prose as it can be watched.
        """
        loaded = self._loaded(ckpt_id)
        mask_id = int(loaded.model.cfg.mask_token_id)
        length = max(1, min(int(length), MAX_LENGTH))
        steps = max(1, min(int(steps), MAX_STEPS))
        prefix = (loaded.tokenizer.encode(prompt, bos=True) if prompt.strip()
                  else [loaded.tokenizer.bos_id])

        t0 = time.monotonic()
        ids, trace = diffusion_generate(
            loaded.model, length=length, steps=steps, prefix=prefix,
            temperature=float(temperature), top_k=int(top_k) or None, top_p=float(top_p),
            remask=remask, seed=seed, device=loaded.device, trace=True)
        elapsed = time.monotonic() - t0

        return {
            "checkpoint": loaded.info.rel,
            "device": loaded.device,
            "prefix_len": len(prefix),
            "text": loaded.tokenizer.decode([i for i in ids if i != mask_id]),
            "steps": [
                {"step": s.step, "cells": self._cells(loaded, s.ids, mask_id),
                 "committed": s.committed,
                 "confidence": [round(c, 3) for c in s.confidence],
                 "remaining": s.remaining}
                for s in trace
            ],
            "elapsed_s": round(elapsed, 3),
            "passes": len(trace) - 1,
            # The headline arithmetic of the paradigm: an autoregressive model would have
            # needed one forward pass per token. Shown as a ratio because that ratio is the
            # only thing diffusion is unambiguously better at.
            "tokens_per_pass": round(length / max(len(trace) - 1, 1), 2),
        }

    def infill(self, ckpt_id: str, prefix: str, suffix: str, length: int = 12,
               steps: int = 12, temperature: float = 0.8,
               seed: int | None = None) -> dict:
        """Write the middle, given both ends — the capability an AR model does not have."""
        loaded = self._loaded(ckpt_id)
        mask_id = int(loaded.model.cfg.mask_token_id)
        if not prefix.strip() and not suffix.strip():
            raise InferError("give it at least one end to work from")
        pre = loaded.tokenizer.encode(prefix, bos=True) if prefix.strip() else [
            loaded.tokenizer.bos_id]
        suf = loaded.tokenizer.encode(suffix) if suffix.strip() else []
        length = max(1, min(int(length), MAX_LENGTH))
        steps = max(1, min(int(steps), MAX_STEPS))
        middle, trace = infill(loaded.model, pre, suf, length=length, steps=steps,
                               temperature=float(temperature), seed=seed,
                               device=loaded.device, trace=True)
        return {
            "checkpoint": loaded.info.rel,
            "device": loaded.device,
            "prefix": prefix, "suffix": suffix,
            "middle": loaded.tokenizer.decode([i for i in middle if i != mask_id]),
            "steps": [
                {"step": s.step, "cells": self._cells(loaded, s.ids, mask_id),
                 "committed": s.committed, "remaining": s.remaining}
                for s in trace
            ],
            "prefix_len": len(pre), "suffix_len": len(suf),
        }

    # ---- measurement -----------------------------------------------------------------------

    def measure(self, ckpt_id: str, kind: str = "elbo", batches: int = 4,
                batch_size: int = 4, buckets: int = 8) -> dict:
        """The validation bound, or cross-entropy against mask rate.

        Bounded much harder than the CLI's defaults: this is a click, and a click that takes
        thirty seconds reads as a broken page. The numbers to put in a write-up come from
        `python -m aksharallm.diffusion <ckpt> elbo`, which the panel says.
        """
        from ..data.loader import TokenDataset

        loaded = self._loaded(ckpt_id)
        val_bin = self._val_bin(loaded)
        batches = max(1, min(int(batches), MAX_BATCHES))
        ds = TokenDataset(str(val_bin), loaded.model.cfg.max_seq_len, loaded.device)

        if kind == "by-t":
            rows = loss_by_t(loaded.model, ds, int(batch_size), batches,
                             buckets=max(2, min(int(buckets), 20)))
            return {"kind": "by-t", "rows": rows, "checkpoint": loaded.info.rel,
                    "device": loaded.device}
        out = elbo(loaded.model, ds, int(batch_size), batches)
        out.update({"kind": "elbo", "checkpoint": loaded.info.rel,
                    "device": loaded.device})
        return out

    def _val_bin(self, loaded) -> Path:
        rel = loaded.info.val_bin
        if not rel:
            raise InferError(
                f"{loaded.info.rel} does not record a validation split, so there is nothing "
                "to measure against.")
        path = Path(rel)
        if not path.is_absolute():
            path = self.root / rel
        if not path.is_file():
            raise InferError(f"the run's validation data ({rel}) is not on disk.")
        return path

    def results(self, limit: int = 20) -> list[dict]:
        """Measurements the CLI has written into `logs/diffusion/`, newest first."""
        import json

        folder = self.root / "logs" / "diffusion"
        if not folder.is_dir():
            return []
        rows = []
        for path in sorted(folder.glob("*.json"), key=lambda p: p.stat().st_mtime,
                           reverse=True)[:limit]:
            try:
                blob = json.loads(path.read_text())
            except (json.JSONDecodeError, OSError):
                continue
            rows.append({"name": path.name, "when": int(path.stat().st_mtime),
                         "checkpoint": blob.get("checkpoint"),
                         "nelbo": blob.get("nelbo"),
                         "kind": "by-t" if "rows" in blob else "elbo"})
        return rows


def sample_rate(batch: int = 1, t_min: float = 1e-3) -> float:
    """One draw of the training mask rate — used by the tab's "surprise me" button."""
    return float(sample_t(batch, t_min, "cpu")[0])
